tests/src/test/scala/whisk/core/containerpool/test/ContainerProxyTests.scala

```diff
@@ -0,0 +1,458 @@
+                initializeCount += 1
+                initPromise.future
+            }
+        }
+        def factory(tid: TransactionId, name: String, exec: Exec, memoryLimit: ByteSize) = Future.successful(container)
+
+        val machine = childActorOf(ContainerProxy.props(factory, ack, store))
+        registerCallback(machine)
+
+        // Start running the action
+        machine ! Run(action, message)
+        expectMsg(Transition(machine, Uninitialized, Running))
+
+        // Schedule the container to be removed
+        machine ! Remove
+
+        // Finish /init, note that /run and log-collecting happens nonetheless
+        initPromise.success(Interval.zero)
+        expectWarmed(invocationNamespace.name, action)
+        expectMsg(Transition(machine, Running, Ready))
+
+        // Remove the container after the transaction finished
+        expectMsg(ContainerRemoved)
+        expectMsg(Transition(machine, Ready, Removing))
+
+        awaitAssert {
+            container.initializeCount shouldBe 1
+            container.runCount shouldBe 1
+            container.logsCount shouldBe 1
+            container.haltCount shouldBe 0 // skips pausing the container
+            container.destroyCount shouldBe 1
+            ack.verify(message.transid, *).repeat(1)
+            store.verify(message.transid, *).repeat(1)
+        }
+    }
+
+    // this tests a Run message in the "Removing" state. The contract between the pool and state-machine
+    // is, that only one Run is to be sent until a "NeedWork" comes back. If we sent a NeedWork but no work is
+    // there, we might run into the final timeout which will schedule a removal of the container. There is a
+    // time window though, in which the pool doesn't know of that decision yet. We handle the collision by
+    // sending the Run back to the pool so it can reschedule.
+    it should "send back a Run message which got sent before the container decided to remove itself" in within(timeout) {
+        val destroyPromise = Promise[Unit]
+        val container = new TestContainer {
+            override def destroy()(implicit transid: TransactionId): Future[Unit] = {
+                destroyCount += 1
+                destroyPromise.future
+            }
+        }
+        def factory(tid: TransactionId, name: String, exec: Exec, memoryLimit: ByteSize) = Future.successful(container)
+
+        val machine = childActorOf(ContainerProxy.props(factory, ack, store))
+        registerCallback(machine)
+        run(machine, Uninitialized)
+        timeout(machine)
+        expectPause(machine)
+        timeout(machine)
+
+        // We don't know of this timeout, so we schedule a run.
+        machine ! Run(action, message)
+
+        // State-machine shuts down nonetheless.
+        expectMsg(ContainerRemoved)
+        expectMsg(Transition(machine, Paused, Removing))
+
+        // Pool gets the message again.
+        expectMsg(Run(action, message))
```

Comment:
wondering if we'll want a prioritized message queue in the future - so that message that were ready to execute aren't delayed at the back of the line through no fault of their own.